src/types/memoryview.rs

```diff
@@ -16,6 +18,58 @@ impl PyMemoryView {
                 .from_owned_ptr_or_err(ffi::PyMemoryView_FromObject(src.as_ptr()))
         }
     }
+
+    /// Creates a new Python memoryview from a raw pointer and length.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Safety
+    ///
+    /// This function dereferences the raw pointer `ptr` as the
+    /// leading pointer of a slice of length `len`. [As with
+    /// `std::slice::from_raw_parts`, this is
+    /// unsafe](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety).
+    pub unsafe fn from_ptr(
+        py: Python<'_>,
+        ptr: *const u8,
+        len: usize,
+        is_writable: bool,
+    ) -> &PyMemoryView {
+        let flags = if is_writable {
+            ffi::PyBUF_WRITE
+        } else {
+            ffi::PyBUF_READ
+        };
+
+        py.from_owned_ptr(ffi::PyMemoryView_FromMemory(
+            ptr as *mut _,
+            len as ffi::Py_ssize_t,
+            flags,
+        ))
+    }
+
+    /// Returns the length of the memoryview.
+    pub fn len(&self) -> usize {
+        let size = unsafe { ffi::Py_SIZE(self.as_ptr()) };
+        // non-negative Py_ssize_t should always fit into Rust usize
+        size as usize
+    }
+
+    /// Checks if the memoryview is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `self` cast as a `PySequence`.
+    pub fn as_sequence(&self) -> &PySequence {
+        unsafe { self.downcast_unchecked() }
+    }
+
+    /// Gets the Python string as a byte slice.
+    #[inline]
+    pub fn as_buffer<T: Element>(&self) -> PyResult<PyBuffer<T>> {
```

Comment:
I'm trying to think, is there a memory safety concern casting from bytes to arbitrary elements `T` here?   Potentially as long as the buffer format is checked, it's fine? We require `Element: Copy` which implies these types are all POD and don't own heap allocations to trigger use-after-free etc

Reply:
If there is, it was already be present in `PyBuffer::get`, wasn't it? And that one does indeed check the buffer format using `Element::is_compatible_format`.

What I would like to see is a unit test to check whether this actually works for anything but `PyBuffer<u8>`.